sjsonnet/src/sjsonnet/Std.scala

```diff
@@ -1130,78 +1039,91 @@ class Std {
     builtinWithDefaults("setUnion", "a" -> null, "b" -> null, "keyF" -> Val.False(dummyPos)) { (args, pos, ev) =>
       val a = args(0) match {
         case arr: Val.Arr => arr.asLazyArray
-        case str: Val.Str => stringChars(pos, str.value).asLazyArray
+        case str: Val.Str if !ev.settings.strictSetOperations => stringChars(pos, str.value).asLazyArray
         case _ => Error.fail("Arguments must be either arrays or strings")
       }
       val b = args(1) match {
         case arr: Val.Arr => arr.asLazyArray
-        case str: Val.Str => stringChars(pos, str.value).asLazyArray
+        case str: Val.Str if !ev.settings.strictSetOperations => stringChars(pos, str.value).asLazyArray
         case _ => Error.fail("Arguments must be either arrays or strings")
       }
-      val concat = new Val.Arr(pos, a ++ b)
-      uniqArr(pos, ev, sortArr(pos, ev, concat, args(2)), args(2))
+      if (ev.settings.strictSetOperations && a.length == 0) {
+        args(0)
+      } else if (ev.settings.strictSetOperations && b.length == 0) {
+        args(1)
+      } else {
+        val concat = new Val.Arr(pos, a ++ b)
+        uniqArr(pos, ev, sortArr(pos, ev, concat, args(2)), args(2))
+      }
     },
-    "setInter" -> SetInter,
-    builtinWithDefaults("setDiff", "a" -> null, "b" -> null, "keyF" -> Val.False(dummyPos)) { (args, pos, ev) =>
+    builtinWithDefaults("setInter", "a" -> null, "b" -> null, "keyF" -> Val.False(dummyPos)) { (args, pos, ev) =>
+      val keyF = args(2)
+      validateSet(ev, pos, keyF, args(0))
+      validateSet(ev, pos, keyF, args(1))
 
       val a = args(0) match {
         case arr: Val.Arr => arr.asLazyArray
-        case str: Val.Str => stringChars(pos, str.value).asLazyArray
+        case str: Val.Str if !ev.settings.strictSetOperations => stringChars(pos, str.value).asLazyArray
         case _ => Error.fail("Arguments must be either arrays or strings")
       }
       val b = args(1) match {
         case arr: Val.Arr => arr.asLazyArray
-        case str: Val.Str => stringChars(pos, str.value).asLazyArray
+        case str: Val.Str if !ev.settings.strictSetOperations => stringChars(pos, str.value).asLazyArray
         case _ => Error.fail("Arguments must be either arrays or strings")
       }
 
+      val out = new mutable.ArrayBuffer[Lazy]
+
+      // The intersection will always be, at most, the size of the smallest set.
+      val sets = if (b.length < a.length) (b, a) else (a, b)
+      for (v <- sets._1) {
+        if (existsInSet(ev, pos, keyF, sets._2, v.force) &&
+          (ev.settings.strictSetOperations || !existsInSet(ev, pos, keyF, out.toArray, v.force))) {
```

Comment:
In non-strict mode, this is going to call `out.toArray` on every element added and that's going to perform a lot of garbage allocation.

Given that we're using an `ArrayBuffer` here, I think we might be able to avoid those repeated array copies by simply passing `out` here and then in `existsInSet` changing it to `arr: mutable.IndexedSeq[Lazy]`.